src/langchain_chatbot/components/langchain_utils_merged.py

```diff
@@ -0,0 +1,147 @@
+
+    def compose_final_prompt(self, instructions, question, table_context):
+        return f"{instructions}\nQuestion: {question}\nContext: {table_context}"
+
+
+
+    @st.cache_resource
+    def get_chain(self, question):
+        print("Creating chain")
+        #question = itemgetter("question")
+        # Retrieve similar documents based on the question
+        table_context = self.document_retriever.find_top_k_similar(question=question, k=4)
+        
+        # Format the context for SQL generation
+        formatted_context = self.generate_final_output(table_context)
+        
+        # Create the final SQL prompt
+        final_prompt_for_sql_generate = self.compose_final_prompt(
+            instructions="Provide an SQL query for the following context:",
+            question=question,
+            table_context=formatted_context
+        )
+        
+        # Generate the SQL query
+        sql_query = self.create_sql_query_chain(final_prompt_for_sql_generate)
+        
+        # Execute SQL and rephrase the answer
+        with self.sql_generator.db.connect() as connection:
+            sql_results = connection.execute(sql_query).fetchall()
+        
+        # Rephrase the final answer
+        final_answer = self.my_rephrase_prompt(question, sql_query, sql_results)
+        return final_answer    
+
+   
+    def my_rephrase_prompt(self, question, sql_query, sql_results): 
+        my_prompt=f""" Given the following user question, corresponding SQL query, and SQL result, answer the user question.
+
+        Question: {question}
+        SQL Query: {sql_query}
+        SQL Result: {sql_results}
+        Answer: 
+        """
+        
+        final_answer_rephrase = self.llm.invoke(my_prompt)
+
+        return final_answer_rephrase
+
+
+    def invoke_chain(self, question, messages):
+        try:
+            chain = self.get_chain()
+            history = create_history(messages)
+            response = chain.invoke({"question": question, "top_k": 3, "messages": history.messages})
+
+            history.add_user_message(question)
+            history.add_ai_message(response)
+
+            if not response or response.strip() == "" or "error" in response:
+                return "Sorry, I couldn't find any specific information related to your query. Please try asking something else or provide more details!"
+
+            return response 
+
+        except Exception as e:
+            print(f"Error invoking chain: {e}")
+            return "Sorry, an error occurred while processing your request."
+
+
+    def create_history(messages):
+        history = ChatMessageHistory()
+        for message in messages:
+            if message["role"] == "user":
+                history.add_user_message(message["content"])
+            else:
+                history.add_ai_message(message["content"])
+        return history
+
+
+#Testing 
+if __name__=="__main__":
+   Orchestrator = Orchestrator(db, llm_provider="openai")
```

Comment:
Orchestrator = Orchestrator(db, llm_provider="openai")
 
 you use same name for type and name .
 
 should be 
 
  orchestrator = Orchestrator(db, llm_provider="openai")
  
  variables lower snake case.
  
  this program working ?